File: src/main/java/net/zhuruoling/omms/central/network/session/handler/builtin/controller/GetControllerStatusRequestHandler.java
package net.zhuruoling.omms.central.network.session.handler.builtin.controller;

import net.zhuruoling.omms.central.controller.ControllerManager;
import net.zhuruoling.omms.central.network.old.session.SessionContext;
import net.zhuruoling.omms.central.network.session.request.Request;
import net.zhuruoling.omms.central.network.session.response.Response;
import net.zhuruoling.omms.central.network.session.response.Result;
import net.zhuruoling.omms.central.network.session.handler.builtin.BuiltinRequestHandler;
import net.zhuruoling.omms.central.permission.Permission;
import net.zhuruoling.omms.central.util.Util;
import org.jetbrains.annotations.NotNull;

public class GetControllerStatusRequestHandler extends BuiltinRequestHandler {
    @Override
    public Response handle(@NotNull Request request, SessionContext session) {
        var controllerId = request.getContent("id");
        if (ControllerManager.INSTANCE.getControllers().containsKey(controllerId)) {
            var status = ControllerManager.INSTANCE.getControllerStatus(kotlin.collections.CollectionsKt.mutableListOf(controllerId));
            if (status.containsKey(controllerId)){
                return new Response().withResponseCode(Result.CONTROLLER_STATUS_GOT).withContentPair("status", Util.toJson(status.get(controllerId)));
            }else {
                return new Response().withResponseCode(Result.CONTROLLER_NO_STATUS);
            }
        } else {
            return new Response().withResponseCode(Result.CONTROLLER_NOT_EXIST).withContentPair("controllerId", controllerId);
        }
    }

    @Override
    public Permission requiresPermission() {
        return null;
    }
}

File: src/main/java/icu/takeneko/omms/central/network/session/handler/builtin/controller/GetControllerStatusRequestHandler.kt
package icu.takeneko.omms.central.network.session.handler.builtin.controller

import icu.takeneko.omms.central.controller.ControllerManager.controllers
import icu.takeneko.omms.central.controller.ControllerManager.getControllerStatus
import icu.takeneko.omms.central.network.session.SessionContext
import icu.takeneko.omms.central.network.session.handler.builtin.BuiltinRequestHandler
import icu.takeneko.omms.central.network.session.request.Request
import icu.takeneko.omms.central.network.session.response.Response
import icu.takeneko.omms.central.permission.Permission
import icu.takeneko.omms.central.util.Util

class GetControllerStatusRequestHandler : BuiltinRequestHandler() {
    override fun handle(request: Request, session: SessionContext): Response {
        val controllerId = request.getContent("id")
        if (controllers.containsKey(controllerId)) {
            val status = getControllerStatus(mutableListOf(controllerId))
            return request.success().withContentPair(
                "status", Util.toJson(
                    status[controllerId]!!
                )
            )
        } else {
            return request.fail().withContentPair("controllerId", controllerId)
        }
    }

    override fun requiresPermission(): Permission? {
        return null
    }
}
